PackageBuilder/Build.cs

```diff
@@ -275,6 +282,109 @@ ListingSource MakeListingSourceFromManifest(VRCPackageManifest manifest)
+
+            // packages will be picked from other repositories are known packages
+            packageIds.UnionWith(listSourceVpmPackages.Keys);
+
+            // packages added to repository from URL or github repository are known packages
+            packageIds.UnionWith(packages.Select(x => x.Id));
+
+            return packageIds;
+        }
+
+        async Task<IEnumerable<VRCPackageManifest>> DownloadPackageManifestFromVpmRepository(string url,
+            List<(string id, VpmPackageInfo info)> packages, HashSet<string> knownPackages)
+        {
+            Serilog.Log.Information($"Downloading from vpm repository {url}");
+            var response = await Http.GetAsync(url);
+            var jsonText = await response.Content.ReadAsStringAsync();
+            var repository = JsonConvert.DeserializeObject<VRCRepoList>(jsonText, Settings.JsonReadOptions);
+
+            var result = new List<VRCPackageManifest>();
+
+            foreach (var (packageId, packageInfo) in packages)
+            {
+                if (!repository.Versions.TryGetValue(packageId, out var versions))
+                {
+                    Serilog.Log.Warning($"{packageId} is not defined in {url}!");
+                    continue;
+                }
+
+                foreach (var (versionString, package) in versions.Versions)
+                {
+                    if (!Version.TryParse(versionString, out var version))
+                    {
+                        Serilog.Log.Warning($"We found invalid version of {packageId} in {url}: {versionString}");
+                        continue;
+                    }
+
+                    if (!packageInfo.includePrerelease && version.IsPreRelease) continue;
+                    // TODO: add option to specify include version range
+
+                    // In Settings.JsonReadOptions, we have JsonConverter that deserializes IVRCPackage with VRCPackageManifest
+                    // so this cast should not be failed.
+                    if (package is not VRCPackageManifest manifest)
+                    {
+                        Serilog.Log.Error($"logic failure: deserialized IVRCPackage is not VRCPackageManifest!");
+                        continue;
+                    }
+
+                    Serilog.Log.Information($"Found {PackageName(manifest)} {manifest.Version}, adding to listing.");
+                    result.Add(manifest);
+
+                    // if there are unknown dependency packages, warn
+                    if (!manifest.VPMDependencies.Keys.All(knownPackages.Contains))
+                    {
+                        // TODO: should this be error and omit from package listing?
+                        Serilog.Log.Warning(
+                            $"We found some missing dependency packages in {packageId} version {version}: " +
+                            string.Join(", ", manifest.Dependencies.Keys.Where(dep => !knownPackages.Contains(dep))));
```

Comment:
I'm sorry to come back again.

I believe we need to join the keys from the VPMDependencies property instead of Dependencies so we can match the right values.

from the package
```
{
  "name": "nadena.dev.modular-avatar",
  "displayName": "Modular Avatar",
  "version": "1.8.2",
  "unity": "2019.4",
  "description": "A suite of tools for assembling your avatar out of reusable components",
  "author": {
    "name": "bd_",
    "email": "bd_@nadena.dev",
    "url": "https://github.com/bdunderscore/modular-avatar"
  },
  "url": "https://m-a.nadena.dev",
  "repo": "https://vpm.nadena.dev/vpm.json",
  "dependencies": {
    "com.unity.nuget.newtonsoft-json": "2.0.0"
  },
  "vpmDependencies": {
    "com.vrchat.avatars": ">=3.2.0",
    "nadena.dev.ndmf": ">=1.2.4 <2.0.0-a"
  }
}
```

I receive le following log entry
`  Warning: We found some missing dependency packages in nadena.dev.modular-avatar version 1.8.2: com.unity.nuget.newtonsoft-json`

Reply:
Oh sorry, Thank you for letting me know.